File: verification/scripts/python/core/test_polynomial.py
# ============================================================================
# Polynomial Regression Test Reference Implementation (Python with statsmodels)
# ============================================================================
# This script generates reference values for polynomial regression using Python's
# statsmodels library. The test validates that the Rust implementation matches
# statsmodels' behavior for coefficient estimates and statistics.
#
# Source: statsmodels.regression.linear_model.OLS with polynomial features
# Reference: "Applied Linear Statistical Models" (Kutner et al.)
#
# Usage:
#   python test_polynomial.py --csv [csv_path] --output-dir [output_dir] [--degree DEGREE]
#   Args:
#     --csv        - Path to CSV file (first col = response, second col = predictor)
#                    Default: ../../datasets/csv/mtcars.csv
#     --output-dir - Path to output directory
#                    Default: ../../results/python
#     --degree     - Polynomial degree (default: 2)
# ============================================================================

import argparse
import numpy as np
import pandas as pd
import statsmodels.api as sm
import json
import sys
from pathlib import Path


class NumpyEncoder(json.JSONEncoder):
    """Helper for JSON serialization of numpy types"""
    def default(self, obj):
        if isinstance(obj, np.integer):
            return int(obj)
        if isinstance(obj, np.floating):
            return float(obj)
        if isinstance(obj, np.ndarray):
            return obj.tolist()
        return super(NumpyEncoder, self).default(obj)


def convert_categorical_to_numeric(data, dataset_name):
    """Convert categorical columns to numeric representations (0-based factor codes)."""
    non_numeric_cols = data.select_dtypes(exclude=[np.number]).columns.tolist()

    if len(non_numeric_cols) > 0:
        print(f"INFO: Dataset '{dataset_name}' contains non-numeric columns: "
              f"{', '.join(non_numeric_cols)}")
        print("Converting categorical variables to numeric factor codes...")

        for col in non_numeric_cols:
            # First try to convert to numeric directly (handles numeric strings)
            numeric_vals = pd.to_numeric(data[col], errors='coerce')
            if not numeric_vals.isna().any():
                # All values converted successfully - use numeric values
                data[col] = numeric_vals
            else:
                # Contains non-numeric strings - convert to factor codes (0-based)
                codes, _ = pd.factorize(data[col])
                data[col] = codes.astype(float)
                print(f"  Column '{col}' encoded as 0-based factor codes")

    return data


def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(
        description="Generate polynomial regression reference values"
    )
    parser.add_argument(
        "--csv",
        default="../../datasets/csv/mtcars.csv",
        help="Path to CSV file (first col = response, second col = predictor)"
    )
    parser.add_argument(
        "--output-dir",
        default="../../results/python",
        help="Path to output directory"
    )
    parser.add_argument(
        "--degree",
        type=int,
        default=2,
        help="Polynomial degree (default: 2)"
    )

    args = parser.parse_args()

    if args.degree < 1:
        print("ERROR: Polynomial degree must be at least 1")
        sys.exit(1)

    # Resolve paths - use absolute paths directly, resolve relative paths from cwd
    csv_path = Path(args.csv) if Path(args.csv).is_absolute() else Path.cwd() / args.csv
    output_dir = Path(args.output_dir) if Path(args.output_dir).is_absolute() else Path.cwd() / args.output_dir

    # Validate CSV path
    if not csv_path.exists():
        print(f"ERROR: CSV file not found: {csv_path}")
        sys.exit(1)

    # Extract dataset name from filename
    dataset_name = csv_path.stem
    degree = args.degree

    print(f"Running polynomial regression (degree={degree}) test on dataset: {dataset_name}")

    # Load data
    data = pd.read_csv(csv_path)
    data = convert_categorical_to_numeric(data, dataset_name)

    # Extract response (first column) and FIRST predictor (second column)
    # Polynomial regression uses a single predictor
    response_col = data.columns[0]
    predictor_col = data.columns[1]

    y = data[response_col].values
    x = data[predictor_col].values

    n = len(y)
    k = degree  # number of slope terms

    if n <= degree + 1:
        print(f"ERROR: Insufficient data: n={n}, degree={degree} (need n > degree+1)")
        sys.exit(1)

    # Build polynomial design matrix: [1, x, x^2, ..., x^degree]
    # This matches our implementation which uses raw (monomial) polynomial basis
    X_poly = np.column_stack([x ** d for d in range(degree + 1)])  # cols: 1, x, x^2, ...

    # Fit OLS model (intercept is already included in X_poly as x^0 = 1)
    model = sm.OLS(y, X_poly)
    results = model.fit()

    # Variable names matching our Rust implementation
    variable_names = ['Intercept'] + [f'x^{d}' for d in range(1, degree + 1)]

    # Coefficients and statistics
    coefficients = results.params.tolist()
    std_errors = results.bse.tolist()
    t_stats = results.tvalues.tolist()
    p_values = results.pvalues.tolist()

    # Model fit statistics
    df_residual = int(results.df_resid)
    df_model = degree

    r_squared = results.rsquared
    adj_r_squared = results.rsquared_adj
    f_statistic = results.fvalue
    f_p_value = results.f_pvalue

    # Residuals and error metrics
    residuals_arr = results.resid
    residuals = residuals_arr.tolist()
    fitted_values = results.fittedvalues.tolist()
    mse = results.mse_resid
    rmse = float(np.sqrt(mse))
    mae = float(np.mean(np.abs(residuals_arr)))
    residual_std_error = float(np.sqrt(mse))

    # Model selection criteria
    log_likelihood = results.llf
    aic_val = results.aic
    bic_val = results.bic

    # Confidence intervals (95%)
    ci = results.conf_int(alpha=0.05)
    conf_int_lower = ci[:, 0].tolist()
    conf_int_upper = ci[:, 1].tolist()

    # Build formula string
    formula_str = f"{response_col} ~ poly({predictor_col}, {degree}, raw=True)"

    # Build result object
    result = {
        "test": "polynomial",
        "method": "statsmodels",
        "dataset": dataset_name,
        "formula": formula_str,
        "degree": degree,
        "n": n,
        "k": k,
        "df_residual": df_residual,
        "df_model": df_model,
        "variable_names": variable_names,
        "coefficients": coefficients,
        "std_errors": std_errors,
        "t_stats": t_stats,
        "p_values": p_values,
        "r_squared": r_squared,
        "adj_r_squared": adj_r_squared,
        "f_statistic": f_statistic,
        "f_p_value": f_p_value,
        "mse": mse,
        "rmse": rmse,
        "mae": mae,
        "residual_std_error": residual_std_error,
        "log_likelihood": log_likelihood,
        "aic": aic_val,
        "bic": bic_val,
        "conf_int_lower": conf_int_lower,
        "conf_int_upper": conf_int_upper,
        "fitted_values": fitted_values,
        "residuals": residuals
    }

    # Ensure output directory exists
    output_dir.mkdir(parents=True, exist_ok=True)

    # Write output
    output_file = output_dir / f"{dataset_name}_polynomial_degree{degree}.json"
    with open(output_file, 'w') as f:
        json.dump(result, f, cls=NumpyEncoder, indent=2)

    print(f"Wrote: {output_file.absolute()}")


if __name__ == "__main__":
    main()

File: verification/scripts/r/core/test_polynomial.R
# ============================================================================
# Polynomial Regression Test Reference Implementation (R with lm)
# ============================================================================
# This script generates reference values for polynomial regression using R's
# lm() function with poly(x, degree, raw=TRUE). The test validates that the
# Rust implementation matches R's behavior for coefficient estimates and
# statistics.
#
# Source: stats::lm with poly(x, degree, raw=TRUE)
# Reference: "Applied Linear Statistical Models" (Kutner et al.)
#
# Usage:
#   Rscript test_polynomial.R [csv_path] [output_dir] [degree]
#   Args:
#     csv_path    - Path to CSV file (first col = response, second col = predictor)
#                   Default: ../../../datasets/csv/mtcars.csv
#     output_dir  - Path to output directory
#                   Default: ../../../results/r
#     degree      - Polynomial degree (default: 2)
# ============================================================================

# Get script directory and resolve relative paths
args_all <- commandArgs(trailingOnly = FALSE)
script_arg <- args_all[grep("^--file=", args_all)]
if (length(script_arg) > 0) {
  script_path <- dirname(sub("--file=", "", script_arg))
  original_wd <- getwd()
  setwd(script_path)
} else {
  original_wd <- getwd()
}

# Add user library to path
user_lib <- file.path(Sys.getenv("USERPROFILE"), "Documents", "R", "win-library", "4.4")
if (dir.exists(user_lib)) {
  .libPaths(c(user_lib, .libPaths()))
}

# Load required packages
if (!require("jsonlite", quietly = TRUE)) {
  stop("Package 'jsonlite' is required. Install with: install.packages('jsonlite')")
}

# Helper function to convert categorical columns to numeric
convert_categorical_to_numeric <- function(data, dataset_name) {
  non_numeric_cols <- names(data)[sapply(data, function(x) !is.numeric(x))]

  if (length(non_numeric_cols) > 0) {
    cat(paste0("INFO: Dataset '", dataset_name, "' contains non-numeric columns: ",
               paste(non_numeric_cols, collapse = ", "), "\n"))
    cat("Converting categorical variables to numeric representations...\n")

    for (col in non_numeric_cols) {
      if (is.factor(data[[col]])) {
        data[[col]] <- as.numeric(data[[col]])
      } else if (is.character(data[[col]])) {
        temp_numeric <- as.numeric(data[[col]])
        if (any(is.na(temp_numeric))) {
          data[[col]] <- as.numeric(as.factor(data[[col]]))
        } else {
          data[[col]] <- temp_numeric
        }
      }
    }
  }

  return(data)
}

# Resolve paths relative to original working directory
resolve_path <- function(path) {
  is_absolute <- grepl("^(/|[A-Za-z]:)", path)
  if (!is_absolute) {
    file.path(original_wd, path)
  } else {
    path
  }
}

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Set defaults
default_csv    <- "../../../datasets/csv/mtcars.csv"
default_output <- "../../../results/r"
default_degree <- 2L

# Parse arguments
csv_path_raw   <- ifelse(length(args) >= 1, args[1], default_csv)
output_dir_raw <- ifelse(length(args) >= 2, args[2], default_output)
degree         <- as.integer(ifelse(length(args) >= 3, args[3], default_degree))

# Resolve paths
csv_path   <- resolve_path(csv_path_raw)
output_dir <- resolve_path(output_dir_raw)

# Validate CSV path
if (!file.exists(csv_path)) {
  stop(paste("CSV file not found:", csv_path))
}

if (degree < 1L) {
  stop("Polynomial degree must be at least 1")
}

# Extract dataset name from filename
dataset_name <- tools::file_path_sans_ext(basename(csv_path))

cat(sprintf("Running polynomial regression (degree=%d) test on dataset: %s\n",
            degree, dataset_name))

# Load data
data <- read.csv(csv_path)
data <- convert_categorical_to_numeric(data, dataset_name)

# Extract response (first column) and FIRST predictor (second column)
# Polynomial regression uses a single predictor
response_col  <- names(data)[1]
predictor_col <- names(data)[2]

y <- data[[response_col]]
x <- data[[predictor_col]]

n <- length(y)
k <- degree  # number of slope terms (no intercept counted)

# Require sufficient data
if (n <= degree + 1) {
  stop(sprintf("Insufficient data: n=%d, degree=%d (need n > degree+1)", n, degree))
}

# Build formula: y ~ poly(x, degree, raw=TRUE)
# raw=TRUE uses monomial basis (1, x, x^2, ...) matching our implementation
formula_str <- sprintf("%s ~ poly(%s, %d, raw=TRUE)", response_col, predictor_col, degree)
formula      <- as.formula(formula_str)

# Fit OLS model with polynomial terms
fit         <- lm(formula, data = data)
summary_fit <- summary(fit)

# Extract coefficients
coefs      <- summary_fit$coefficients
coefficients <- as.vector(coefs[, 1])   # Estimates
std_errors   <- as.vector(coefs[, 2])   # Standard errors
t_stats      <- as.vector(coefs[, 3])   # t-statistics
p_values     <- as.vector(coefs[, 4])   # p-values

# Variable names: rename poly(...) terms to match our output
variable_names <- c("Intercept", paste0("x^", seq_len(degree)))

# Model fit statistics
df_residual <- fit$df.residual
df_model    <- degree

r_squared     <- summary_fit$r.squared
adj_r_squared <- summary_fit$adj.r.squared
f_statistic   <- summary_fit$fstatistic[1]
f_p_value     <- pf(f_statistic, df_model, df_residual, lower.tail = FALSE)

# Residuals and error metrics
residuals_vec  <- as.vector(residuals(fit))
fitted_vals    <- as.vector(fitted(fit))
mse            <- sum(residuals_vec^2) / df_residual
rmse           <- sqrt(mse)
mae            <- mean(abs(residuals_vec))
residual_std_error <- summary_fit$sigma

# Model selection criteria
log_likelihood <- as.numeric(logLik(fit))
aic_val        <- AIC(fit)
bic_val        <- BIC(fit)

# Confidence intervals (95%)
ci             <- confint(fit, level = 0.95)
conf_int_lower <- as.vector(ci[, 1])
conf_int_upper <- as.vector(ci[, 2])

# Build result object
result <- list(
  test       = "polynomial",
  method     = "lm",
  dataset    = dataset_name,
  formula    = formula_str,
  degree     = degree,
  n          = n,
  k          = k,
  df_residual    = as.integer(df_residual),
  df_model       = as.integer(df_model),
  variable_names = variable_names,
  coefficients   = coefficients,
  std_errors     = std_errors,
  t_stats        = t_stats,
  p_values       = p_values,
  r_squared      = r_squared,
  adj_r_squared  = adj_r_squared,
  f_statistic    = f_statistic,
  f_p_value      = f_p_value,
  mse            = mse,
  rmse           = rmse,
  mae            = mae,
  residual_std_error = residual_std_error,
  log_likelihood = log_likelihood,
  aic            = aic_val,
  bic            = bic_val,
  conf_int_lower = conf_int_lower,
  conf_int_upper = conf_int_upper,
  fitted_values  = fitted_vals,
  residuals      = residuals_vec
)

# Ensure output directory exists
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

# Write output
output_file <- file.path(output_dir, paste0(dataset_name, "_polynomial_degree", degree, ".json"))
write_json(result, output_file, pretty = TRUE, auto_unbox = TRUE, digits = 22)

cat(sprintf("Wrote: %s\n", normalizePath(output_file)))
